model/src/main/kotlin/utils/FindingsMatcher.kt

```diff
@@ -129,68 +111,101 @@ class FindingsMatcher(
     private fun matchFileFindings(
         licenses: List<LicenseFinding>,
         copyrights: List<CopyrightFinding>
-    ): Map<String, Set<CopyrightFinding>> {
+    ): Map<LicenseFinding, Set<CopyrightFinding>> {
         require((licenses.map { it.location.path } + copyrights.map { it.location.path }).distinct().size <= 1) {
             "The given license and copyright findings must all point to the same file."
         }
 
         // If there is only a single license finding, associate all copyright findings with that license. If there is
         // no license return no matches.
-        if (licenses.size <= 1) return licenses.associateBy({ it.license }, { copyrights.toSet() })
+        if (licenses.size <= 1) return licenses.associateWith { copyrights.toSet() }
 
         // If there are multiple license findings in a single file, search for the closest copyright statements
         // for each of these, if any.
-        val result = mutableMapOf<String, MutableSet<CopyrightFinding>>()
-        licenses.forEach { (license, location) ->
-            val closestCopyrights = getClosestCopyrightStatements(copyrights, location.startLine, location.endLine)
-            result.getOrPut(license) { mutableSetOf() } += closestCopyrights
+        return licenses.associateWith { licenseFinding ->
+            getClosestCopyrightStatements(
+                copyrights,
+                licenseFinding.location.startLine,
+                licenseFinding.location.endLine
+            )
         }
-
-        return result
     }
 
     /**
-     * Return an association of the given [copyrightFindings] to [licenseFindings].
-     * Copyright findings are either matched to a license finding located nearby in the same file or to a license
-     * finding pointing to a license file. Whether a file is a license file is determined by the
-     * [FileMatcher] passed to the constructor. All [CopyrightFindings]s which cannot be matched are not present
-     * in the result while all given [licenseFindings] are contained in the result exactly once.
+     * Associate the [copyrightFindings] to the [licenseFindings]. Copyright findings are matched to license findings
+     * located nearby in the same file. Copyright findings that are not located close to a license finding are
+     * associated to the root licenses instead. The root licenses are the licenses found in any of the license files
+     * defined by [licenseFileMatcher].
      */
-    fun match(licenseFindings: Collection<LicenseFinding>, copyrightFindings: Collection<CopyrightFinding>):
-            Set<LicenseFindings> {
+    fun matchFindings(
+        licenseFindings: Set<LicenseFinding>,
+        copyrightFindings: Set<CopyrightFinding>
```

Comment:
Nit: These inputs could also be generic collections.

Reply:
Yes, but I wonder if we need the more generic API, when working with findings it's never useful to have duplicates. In most cases the input should already be sets, and if I change that I either have to always call `toSet()` which also copies the collection if it is already a set, or add something like `takeIf { it is Set } ?: ...toSet()`.